Update to use the microgrid API v0.18

llucax opened the pull request:
This PR migrates the SDK from the old `frequenz-client-microgrid` using the old microgrid API v0.15 to the new using the API v0.18. The main changes involve updating component models, metrics, and data fetching mechanisms to use the new client library structure.

llucax committed:
Bump dependency for frequenz-client-microgrid to v0.18.0

This introduces the new microgrid API v0.18. We also need to bump the
frequenz-client-common dependency to at least v0.3.6, so it can work
with the new microgrid client.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Remove test for grid connection without a fuse

The microgrid API now doesn't support not reporting a rated fuse for a
grid connection, so we don't need to test for that case anymore.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Skip battery pool bounds tests

The battery pool bounds calculation is buggy and these tests are wrong
(see https://github.com/frequenz-floss/frequenz-sdk-python/issues/1180).
By switching to using ranges of bounds, the buggy behaviour changes and
make these tests fail.

Fixing them is difficult without switching to using ranges natively
first, so we just skip these tests for now.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update release notes

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented on src/frequenz/sdk/microgrid/_old_component_data.py:
```diff
@@ -0,0 +1,1287 @@
+            raise ValueError("No metrics in the samples.")
+
+        # FIXME: This might not be true forever, but the service for now seems to send
+        # all metrics with the same timestamp for now.
```
I think this one can be removed, at least IIRC @tiyash-basu-frequenz confirmed this was going to be like this for the time being. In any case, this is a temporary transitional module, at some point we should start using the telemetry messages as they come from the API.

llucax replied:
Marking as resolved as this is confirmed.

llucax commented on src/frequenz/sdk/microgrid/_old_component_data.py:
```diff
@@ -0,0 +1,1287 @@
+        if not samples.states:
+            return class_(component_id=samples.component_id, timestamp=timestamp)
+
+        # FIXME: Maybe we can have more than one, in which case we need to merge them?
```
I think @tiyash-basu-frequenz also confirmed this was fine for now, but I don't remember if the idea is that more than one could come in the future.

shsms replied:
I think for EV chargers, we will have component state and cable state, both of which are necessary for its operation.  But I guess @tiyash-basu-frequenz means that they will both be inside `states[0].states`, which is also a list.  We will only not have multiple sets of states. 

In that case, we just need to update the warning message below.

llucax replied:
Yeah, we can still have multiple `samples.states[0].states` with the current code, I was referring to multiple.... Mmmm, this is getting confusing having 2 states that are actually different things, I was referring to multiple `samples.states`.

llucax replied:
I got a confirmation that only one state message will come in a telemetry message (sample).

llucax commented on src/frequenz/sdk/microgrid/_old_component_data.py:
```diff
@@ -0,0 +1,1287 @@
+    # we used the API v0.15, as we accessed the fields without checking if the fields
+    # really existed, so the defaul protobuf value of 0.0 for floats was used.
+    # We might beed to review this and if they are not present interpret them as None
+    # instea.
```
This one is tricky, I guess we can leave it as it is now and will have to come back to it forcibly when removing this temporary adap**ta**tion layer.

shsms replied:
```suggestion
    # FIXME: All of this have now a default of 0.0 because this is what it was doing when
    # we used the API v0.15, as we accessed the fields without checking if the fields
    # really existed, so the default protobuf value of 0.0 for floats was used.
    # We might need to review this and if they are not present interpret them as None
    # instead.
```

Also, it is the "adap**ta**tion" layer.

llucax replied:
No comment on the original comment? :D

shsms replied:
oops, I agree, if a field is not there, it should be None, but can be done separately.

llucax replied:
I left the comment in the module, we'll have to address some comments there when removing the adaptation layer module anyway.

llucax commented on src/frequenz/sdk/microgrid/_old_component_data.py:
```diff
@@ -0,0 +1,1287 @@
+                        # Update power bounds from the SOC metric bounds,
+                        # FIXME: We assume only one range is present
+                        # If one bound is None, we assume 0 to match the previous
+                        # behavior of v0.15, but this should eventually be fixed
```
I guess we'll also fix this when removing this adaption layer, not sure if it is worth adding a separate issue for this one.

llucax commented on src/frequenz/sdk/microgrid/_old_component_data.py:
I guess we can disable pylint `fixme` check for this entire module and keep the FIXMEs as they are now to remember to address these issues when removing this adaption layer.

llucax commented on src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py:
```diff
@@ -416,25 +422,32 @@ def _is_battery_state_correct(self, msg: BatteryData) -> bool:
+        # one state, and now it has a set of states, for now we are considering the
+        # component to be working if any of its states is valid. Maybe we need to switch
+        # to the negative and consider a component not working if is in any known
+        # invalid state instead.
```
@shsms Maybe you have a more clear idea of what to do here. Maybe for a first step we can merge as is and create a separate issue to address it in the future?

shsms replied:
We'll have to assume the component is broken.

llucax replied:
<del>In which case? :thinking: </del>

Oh, from the following comment I guess you mean if there is a contradictory state.

llucax commented on src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py:
```diff
@@ -394,14 +394,20 @@ def _is_inverter_state_correct(self, msg: InverterData) -> bool:
+        # one state, and now it has a set of states, for now we are considering the
+        # component to be working if any of its states is valid. Maybe we need to switch
+        # to the negative and consider a component not working if is in any known
+        # invalid state instead.
```
@shsms Maybe you have a more clear idea of what to do here. Maybe for a first step we can merge as is and create a separate issue to address it in the future?

shsms replied:
Sounds unlikely, unless there's a bug, buf if there are multiple contradictory states for a component for the same timestamp, then we'll have to assume the component is broken.

llucax replied:
OK, I will leave it as it is then. I guess checking for inconsistencies will be something new and probably out of the scope of this PR.

llucax commented on src/frequenz/sdk/microgrid/_old_component_data.py:
```diff
@@ -0,0 +1,1287 @@
+        #     EVChargerCableState.EV_LOCKED,
+        #     EVChargerCableState.EV_PLUGGED,
+        # )
+        # TODO: Verify this logic is correct
```
Maybe you can check this @shsms ?

shsms replied:
the only issue in the logic below is that we're keeping only the latest state - it could be the component state of READY/ERROR,  or cable state of PLUGGED/LOCKED etc.

shsms replied:
just a comment about the cable states, written below.

llucax commented on src/frequenz/sdk/microgrid/_power_distributing/_component_managers/_battery_manager.py:
```diff
@@ -84,9 +80,10 @@ def _get_battery_inverter_mappings(
+            component.id
             for component in component_graph.predecessors(battery_id)
-            if component.category == ComponentCategory.INVERTER
+            # TODO: Shouldn't this be (SolarInverter, HybridInverter)?
```
@shsms ? Maybe it is better to create a separate issue anyways as this was a pre-existing issue.

shsms replied:
We were not doing that because there were still a lot of edge configs from the early days that didn't specify the type of the inverter.  Maybe things have changed now and we can start with a log message if the inverter type is not set.

llucax replied:
Yeah, in the current client this should be logged as a warning as soon it is received by the client, and a `UnspecifiedInverter` type will be returned (with `category` `INVERTER` and `type` `UNSPECIFIED`, which is the default when no type is present).

So I guess you suggest to leave it as it is for now, and make it more specific when we confirm we have no inverters without a type anymore.

llucax replied:
Also I guess it would be a topological error to have a non matching inverter type as a predecessor, but we should still try to use it in case it is a configuration error?

llucax replied:
According to @tiyash-basu-frequenz everything should be properly configured by now.

llucax replied:
In any case, I will keep the old behavior in this PR, as it is only about the microgrid API version upgrade.

* https://github.com/frequenz-floss/frequenz-sdk-python/issues/1285

llucax commented on src/frequenz/sdk/microgrid/_power_distributing/_component_status/_battery_status_tracker.py:
```diff
@@ -371,15 +371,15 @@ def _no_critical_error(self, msg: BatteryData | InverterData) -> bool:
-        critical_err = next((err for err in msg.errors if err.level == critical), None)
-        if critical_err is not None:
+        # TODO: This was using ErrorLevel.CRITICAL to see if an error was critical or
+        # warning. I guess now we use the separate errors and warnings fields.
```
Can you have a look at this one too @shsms ?

shsms replied:
This makes sense I think.

llucax commented on src/frequenz/sdk/microgrid/component_graph.py:
```diff
@@ -539,22 +563,63 @@ def refresh_from(
+            issues.append(
+                "Component with ID 0 should be a GridConnectionPoint, "
+                f"but it is a {component!r} instead."
+            )
```
I think this check is safe to remove. I only found a previous check in an unused method `_correct_graph_errors`, removed in 635914ca5683b0195d3c8c0f7af9a2e8a4250298. I think I wrote this code before committing that removal.

FYI @tiyash-basu-frequenz @shsms @ela-kotulska-frequenz, I will remove this check unless someone complains.

llucax replied:
Removing.

llucax commented on src/frequenz/sdk/timeseries/battery_pool/_methods.py:
```diff
@@ -253,6 +253,9 @@ async def _send_on_update(self, min_update_interval: timedelta) -> None:
                 if result is None:
+                    # TODO: This code is unreacheable, as T can't be None. We need to
+                    # figure out which case this was supposed to handle. Maybe we need
+                    # some sentinel value or to allow None in T.
```
@shsms maybe you can check this one?

shsms replied:
looks like it stopped being `T | None` here: https://github.com/frequenz-floss/frequenz-sdk-python/pull/692/commits/bf22d6d669b8570b2a90217c96d38c9bb2813b14

The handling for this case can go away as well.

llucax replied:
Great, thanks for checking!

llucax reviewed (commented):
I think these are all the TODOs/FIXMEs I wrote when doing the update. It would be good to get some feedback.

shsms commented on src/frequenz/sdk/microgrid/_old_component_data.py:
```diff
@@ -0,0 +1,1287 @@
+            ValueError: if the given id is unknown or has a different type.
+        """
+        # pylint: disable-next=import-outside-toplevel,cyclic-import
+        from frequenz.sdk import microgrid
```
why not relative path instead: `from .. import microgrid`

llucax replied:
I guess no reason.
```suggestion
        from .. import microgrid
```

shsms commented on src/frequenz/sdk/microgrid/_old_component_data.py:
```diff
@@ -0,0 +1,1287 @@
+                        sample,
+                    )
+
+        self.active_power_per_phase = cast(PhaseTuple, tuple(active_power_per_phase))
```
wow, amazing this doesn't happen automatically, given that it is just an alias.

llucax replied:
I guess it is because there is a dynamic conversion going on `active_power_per_phase` is a list[float]` and it is converted to a tuple dynamically, so this should have type `tuple[float, ...]` (if the mapping `list[float]` -> tuple[float, ...]` even works), not `tuple[float, float, float]`.

shsms commented on src/frequenz/sdk/microgrid/_old_component_data.py:
```diff
@@ -0,0 +1,1287 @@
+                    ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_EV,
+                    ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_STATION,
+                }
+                & self.states
```
we probably need to drop the `*_AT_STATION` codes.  Cable connected to the station doesn't mean the other end of the cable has a car attached.

llucax replied:
But this is checking that is connected at both ends, the EV and the station. it seems like the new API is splitting the `LOCKED` and `PLUGGED` into both ends, so the intention here was that we make sure both ends are connected to return true for `is_ev_connected`.

@tiyash-basu-frequenz maybe can also help here.

llucax replied:
Thinking a bit more maybe you mean it is redundant because ia EV charging station can only know the EV end is connected if the station end is also connected? Also I was thinking maybe I should add more variables to make it more readable:

```py
has_errors = ComponentStateCode.ERROR not in self.states
is_authorized = (
                ComponentErrorCode.UNAUTHORIZED not in self.errors
                and ComponentErrorCode.UNAUTHORIZED not in self.warnings
            )
both_ends_are locked = bool(
                {
                    ComponentStateCode.EV_CHARGING_CABLE_LOCKED_AT_EV,
                    ComponentStateCode.EV_CHARGING_CABLE_LOCKED_AT_STATION,
                }
                & self.states)
both_ends_are_plugged = bool(
                {
                    ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_EV,
                    ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_STATION,
                }
                & self.states)
return not has_errors and is_authorized and (both_ends_are_locked or both_ends_are_plugged)
```
This also makes me think that I'm not sure if a combination of "plugged at EV-locked at station" should also not be considered as "connected".

llucax replied:
After talking to @tiyash-basu-frequenz I will update to:

```py
    def is_ev_connected(self) -> bool:
        """Check whether an EV is connected to the charger.

        Returns:
            When the charger is not in an error state, whether an EV is connected to
                the charger.
        """
        has_error = ComponentStateCode.ERROR not in self.states
        is_authorized = (
            ComponentErrorCode.UNAUTHORIZED not in self.errors
            and ComponentErrorCode.UNAUTHORIZED not in self.warnings
        )
        is_connected_at_ev = bool(
            {
                ComponentStateCode.EV_CHARGING_CABLE_LOCKED_AT_EV,
                ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_EV,
            }
            & self.states
        )
        is_connected_at_station = bool(
            {
                ComponentStateCode.EV_CHARGING_CABLE_LOCKED_AT_STATION,
                ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_STATION,
            }
            & self.states
        )

        return (
            not has_error
            and is_authorized
            and is_connected_at_ev
            and is_connected_at_station
        )
```

shsms replied:
There's likely only one cable-state.  So it would either say LOCKED_AT_STATION (meaning there's cable, but no car) or LOCKED_AT_EV (meaning there's car, and therefore cable).

So we can't use AT_STATION states for this check.

llucax replied:
Again, question for @tiyash-basu-frequenz as what I understood from the meeting with him is in v0.18 we should check for both.

shsms replied:
ok, in that case, the new and and and function looks good.

shsms commented on src/frequenz/sdk/microgrid/_power_distributing/_component_status/_ev_charger_status_tracker.py:
```diff
@@ -87,13 +84,25 @@ def start(self) -> None:
+                    ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_EV,
+                    ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_STATION,
+                }
+                & ev_data.states
```
Here too, we have to drop the `AT_STATION` states.  An ev charger can't be assigned power if it doesn't have an EV attached, and that's what the `is_working` method is really about.

llucax replied:
Wouldn't it make more sense in this case to just change to something like:

```py
is_connected = ev_data.is_ev_connected()
is_ok = bool({
                ComponentStateCode.READY,
                ComponentStateCode.CHARGING,
                ComponentStateCode.DISCHARGING,
            }
            & ev_data.states)
return is_connected and is_ok
```

?

llucax replied:
I will go for this, @shsms let me know if you think it is wrong.

shsms replied:
is_ev_connected needs a change, but otherwise this looks good.

shsms reviewed (commented):
Only read a little bit of the big commit

shsms commented on tests/microgrid/power_distributing/_component_status/test_ev_charger_status.py:
```diff
@@ -60,9 +60,12 @@ async def test_status_changes(self, mocker: MockerFixture) -> None:
                     active_power=0.0,
-                    component_state=EVChargerComponentState.READY,
-                    cable_state=EVChargerCableState.EV_PLUGGED,
-                )
+                    states={
+                        ComponentStateCode.READY,
+                        ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_EV,
+                        ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_STATION,
+                    },
+                ).to_samples()
```
cable plugged at EV necessarily means cable LOCKED at station.

And it was not the case with the service to send two states for the cable, and I don't think it needs to change now, it is already obvious, so we'd just be sending redundant information.

llucax replied:
Apparently it already changed, at least this is what I understood from a meeting with @tiyash-basu-frequenz

shsms commented on tests/microgrid/power_distributing/_component_status/test_ev_charger_status.py:
```diff
@@ -121,9 +132,12 @@ async def test_status_changes(self, mocker: MockerFixture) -> None:
                     _EV_CHARGER_ID,
                     datetime.now(tz=timezone.utc),
                     active_power=0.0,
-                    component_state=EVChargerComponentState.READY,
-                    cable_state=EVChargerCableState.EV_LOCKED,
-                )
+                    states={
+                        ComponentStateCode.READY,
+                        ComponentStateCode.EV_CHARGING_CABLE_PLUGGED_AT_EV,
+                        ComponentStateCode.EV_CHARGING_CABLE_LOCKED_AT_STATION,
+                    },
```
this is not an exact match.

Old says EV LOCKED.  new says EV PLUGGED.

llucax replied:
Oh, good catch! Will fix.

llucax replied:
The test still passed because actually all permutations are considered as *connected* I guess.

llucax replied:
(fixed anyway)

llucax committed:
Migrate metric identifiers from ComponentMetricId to Metric

This change updates the SDK to use the new `Metric` enum from
`frequenz.client.microgrid.metrics` (and the new `TransitionalMetric`)
in place of the old `ComponentMetricId`.

It adapts internal types and mappings to accept `Metric
| TransitionalMetric` where appropriate.

Imports, request/channel naming, and data-extraction maps across the
microgrid data sourcing and timeseries modules (battery pool, voltage/
frequency streamers, formula engine, etc.) have been updated to the new
metric names (for example `ACTIVE_POWER` → `AC_ACTIVE_POWER`, etc.), and
a `TransitionalMetric` shim is used to preserve compatibility for
metrics that are still referenced by older code paths.

Tests and benchmarks were updated accordingly.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

shsms reviewed (commented):
Still got a number of files to go through.

llucax committed:
Add some missing `@override`

This will help us to make sure all sub-classes are properly updated when
doing changes to method signatures.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Add old component data wrappers

The new API doesn't return component data packets with one timestamp and
several metrics anymore, so we write an adaptation layer to convert
individual metrics to `ComponentData` wrappers so we don't need to
update every location where streaming data is being consumed.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Use new API v0.18 methods

For non data-streaming methods, update to use the new client method
names. For data-streaming methods, use the new adaptation functions in
`_old_component_data.py`.

Because `set_component_power_active()` can now return a the time when
the command expires, we also need to update the tests we use to run
them.

Note that the data-streaming method in the new client version is not
async (it returns a Receiver synchronously) so `await`s are also
removed.

For the `ComponentMetricFetcher` we also remove the `async` for the
wrapper `_subscribe()` methods.

The mock microgrid client was mostly rewritten to adjust it to the new
unified data streaming method.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update `.component_id` -> `.id`

The new `Component` class renamed `component_id` with `id`, so we need
to update all our references.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Rename some "api" to "client" or "api_client"

API is too generic, a more specific name makes the code more clear.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
I'm working on splitting the big WIP commit, so maybe you want to wait for that before you continue with the review.

llucax committed:
Use the new `MicrogridInfo` object

The name in the microgrid API will be changed again to `Microgrid` in
the near future, so internally we just use `microgrid` instead of
`microgrid_info`.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Replace uses of `ComponentCategory` with the new types

The new API client provides a mapping of `ComponentCategory` and
`XxxType`s to Python classes. This commit updates most uses of component
categories with these classes.

This also includes some updates of imports for `Component` and
`ComponentCategory` and some changes on specific component type, like
`GridConnectionPoint`'s metadata.

Finally, the component graph interface to get components was also
updated to use types instead of categories, for which the API was
slightly changed to better represent the filtering criteria. The
keyword for arguments were changed from `component_ids` to
`matching_ids` and `component_categories` to `matching_types`.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Allow for unknown `ComponentCategory`s

The new API client can return also an `int` as a component category
when the client doesn't know the category number.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Improve component graph's filtering

While we are changing the interface of component retrieval in the
component graph, we take the opportunity to make its interface more
generic.

We now accept any Iterable for `matching_ids` and `matching_types`, as
well as one individual items. Most uses of it filter by a single item
anyway so it is annoying to have to wrap it in a Iterable.

Most uses of those methods were updated to use the single-item overload.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Update `Connection` to `ComponentConnection`

To match the new component graph filtering arguments for `components()`,
the `connections()` method now also have its argument names updated:

* `start` -> `matching_sources`
* `end` -> `matching_destinations`

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Allow passing single elements to component graph `connections()`

Like with component graph `components()`, now we support passing a
single element to `matching_sources` and `matching_destinations`.

The `components()` function is also simplified to make filtering more
compact.

Also updates uses with a single element to remove the unnecessary
container.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Replace old `Component`s with specific classes in tests

Many tests build mock components, so now this components need to be
created as concrete subclasses of `Component`.

Some test also has some minor readability improvements.

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax committed:
Use new component states and state snapshots

The new state snapshots in the v0.18 client provide errors and warnings
using a different structure, so we need to account for that.

State code were also unified and there is no more specific enums for
cable state and other component-specific states, all states are part of
the `ComponentStateCode` enum and the component data can have many
state codes in it, while before it contained only one.

In previous versions EV chargers only used one `PLUGGED` or `LOCKED`
status but now the cable status was split in 2 connector ends: station
and EV, so now all combinations of those need to be accounted for. This
might change again on the API side as it ended up a bit messy.

Also the old component data wrappers now can't be frozen because the
`ComponentData` class is not frozen (this is to allow parsing from the
new telemetry messages easier for phases).

Signed-off-by: Leandro Lucarella <luca-frequenz@llucax.com>

llucax commented:
OK, this should be ready for a final review, I split the big WIP commit into smaller commits. The commit split is still not the best, but it should be much more reviewable now. I left a few unrelated changes/improvements (like the removal of unreachable code) for a follow-up PR with just cleanups/improvements that are not strictly part of the v0.18 update (although there still a few unrelated commits in this PR too).

llucax commented on src/frequenz/sdk/microgrid/_old_component_data.py:
```diff
@@ -0,0 +1,1290 @@
+        )
+
+
+def _bound_ranges_to_inclusion_exclusion(
+    bounds: list[Bounds], name: str, sample: MetricSample
+) -> tuple[float, float, float, float]:
```
I have no idea what do you mean by this, but in any case `mypy` doesn't allow us to use implicit returns anyways.

llucax commented on src/frequenz/sdk/microgrid/_power_distributing/_component_managers/_battery_manager.py:
```diff
@@ -8,21 +8,17 @@
-    OperationOutOfRange,
-)
+from frequenz.client.microgrid import ApiClientError, OperationOutOfRange
+from frequenz.client.microgrid.component import Battery, Inverter
```
BS

shsms commented on pyproject.toml:
```diff
@@ -29,8 +29,9 @@ dependencies = [
-  "frequenz-client-microgrid >= 0.9.0, < 0.10.0",
-  "frequenz-client-common >= 0.3.2, < 0.4.0",
+  #"frequenz-client-microgrid >= 0.18.0, < 0.19.0",
+  "frequenz-client-microgrid @ git+https://github.com/frequenz-floss/frequenz-client-microgrid-python@refs/heads/v0.18.x",
```
Is there a PR for the client already?  Can't find one.

llucax commented:
Updated the dependency, I think it would be best to merge when the review is ready, so marking as ready. I just realized I also need to update the release notes.

shsms commented:
Super, you can remove WIP from the commit message as well.  I will continue reviewing now.

shsms reviewed (dismissed):
🎉

llucax commented:
Whooohooo! Rebased and enabled auto-merge.

llucax commented:
OK, I needed to [adapt a new test](https://github.com/frequenz-floss/frequenz-sdk-python/compare/578863929a8bb9381692cd0e48cb0bd1f0bdc266..683bcb24695c9103c090841686476bd5bff3a5be), so it will need a new approval.
